fix(fastsim): fast sim fails when using InstrumentContext.pair_with function

amitlissack opened the pull request:
# Overview

The paired pipette feature was completely neglected by the fast simulation refactor. This PR aims to remedy that by having `PairedInstrumentContext` initialized with an `AbstractPairedInstrument`. For normal cases, we will use `PairedInstrument` just like now. But for fast simulation, we will use an `AbstractPairedInstrument` that does not communicate with the hardware controller at all.

This is a way bigger PR than I'd hoped. But I couldn't think of an easier solution.

closes #8222 

# Changelog

- Create `AbstractPairedInstrument` 
- Have `PairedInstrument` inherit from `AbstractPairedInstrument`
- Create `AbstractInstrument.pair_with` in order to specialize in `InstrumentContextSimulation`.
- `AbstractPairedInstrument` is supplied to `PairedInstrumentContext`'s constructor.
- Created `PairedInstrumentSimulation` for fast simulation
- Moved a lot of location validation out of `PairedInstrument` into `PairedInstrumentContext` so the validation happens for  fastsim too.
- Added tests to validate that same errors are raised in simulation and in fast simulation.
- Fixed the flag unsafe move checking in move_to

# Review requests

Does pipette pairing still work?

I tested by comparing the simulator outputs of user supplied protocol using fast and slow simulation.  The only differences were some of the timings.

# Risk assessment

High in that it impacts a documented feature of our protocol api. **It is however a very rarely used feature.**

Commit:
ABC

Commit:
use implementation

Commit:
tests pass

Commit:
clean up logging in protocol api

Commit:
fixed.... or is it?

DerekMaggio commented:
Can you provide a summary of what changes you made and why you made them? 
I took a look at the code and I see a ton of changes but I am not sure what part/s fixed the issue.

amitlissack commented:
> Can you provide a summary of what changes you made and why you made them?
> I took a look at the code and I see a ton of changes but I am not sure what part/s fixed the issue.

It's still in draft and labeled WIP. Not ready to be looked at.

DerekMaggio commented:
> > Can you provide a summary of what changes you made and why you made them?
> > I took a look at the code and I see a ton of changes but I am not sure what part/s fixed the issue.
> 
> It's still in draft and labeled WIP. Not ready to be looked at.

Oh. Duh. Ignore me

Commit:
Move location processing out of implementation and into the context

Commit:
simulate move_to

Commit:
add tests for paired instrument simulation raising same errors as paired instrument.
needed to move thermocycler flag unsafe moves to thermocycler geometry to be accessed in paired isntrument.

Commit:
lint

Laura-Danielle commented on api/src/opentrons/protocol_api/paired_instrument_context.py:
```diff
@@ -75,16 +74,9 @@ def __init__(
-            hardware_manager,
-            self._log,
-        )
+        self.paired_instrument_obj = implementation
```
Do we want to make this slightly more consistent with the regular protocol context and call this attribute `self._implementation`?

Laura-Danielle reviewed (approved):
One little suggestion that you can ignore. The refactor looks great! Thanks for doing this.

We should just make sure QA does a quick runthrough protocol on a robot.

DerekMaggio commented on api/src/opentrons/protocol_api/paired_instrument_context.py:
```diff
@@ -425,34 +417,42 @@ def aspirate(
         )
 
-        loc: Optional[types.Location] = None
+        # checked_location: Optional[types.Location] = None
```
Should this commented out line be removed?

Commit:
address PR comments.

nusrat813 reviewed (approved):
tested simulation, lgtm 
@Laura-Danielle tested twinning protocol on her robot
